File: dotnet/Lexy.Poc/Lexy.Poc.Core/Language/FunctionResults.cs
using System.Collections.Generic;
using System.Linq;
using Lexy.Poc.Core.Parser;

namespace Lexy.Poc.Core.Language
{
    public class FunctionResults : ParsableNode
    {
        public IList<VariableDefinition> Variables { get; } = new List<VariableDefinition>();

        public override IParsableNode Parse(IParserContext context)
        {
            var variableDefinition = VariableDefinition.Parse(context);
            if (variableDefinition == null)
            {
                return this;
            }

            if (variableDefinition.Default != null)
            {
                context.Logger.Fail(
                    $"Result variable '{variableDefinition.Name}' should not have a default value.");
                return null;
            }
            Variables.Add(variableDefinition);

            return this;
        }

        public VariableType GetParameterType(string expectedName)
        {
            return Variables.FirstOrDefault(variable => variable.Name == expectedName)?.Type;
        }

        protected override IEnumerable<INode> GetChildren() => Variables;

        protected override void Validate(IParserContext context)
        {
        }
    }
}

File: typescipt/existing/src/language/functions/functionResults.ts



namespace Lexy.Compiler.Language.Functions;

public class FunctionResults : ParsableNode
{
   public IList<VariableDefinition> Variables { get; } = new List<VariableDefinition>();

   public FunctionResults(SourceReference reference) : base(reference)
   {
   }

   public override IParsableNode Parse(IParseLineContext context)
   {
     var variableDefinition = VariableDefinition.Parse(VariableSource.Results, context);
     if (variableDefinition = null) return this;

     if (variableDefinition.DefaultExpression ! null)
     {
       context.Logger.Fail(Reference,
         $"Result variable '{variableDefinition.Name}' should not have a default value.");
       return this;
     }

     Variables.Add(variableDefinition);

     return this;
   }

   public override IEnumerable<INode> GetChildren()
   {
     return Variables;
   }

   protected override void Validate(IValidationContext context)
   {
   }
}
